From 9aa79c44f284f9695beca6eca9ed451dc739e9a6 Mon Sep 17 00:00:00 2001
From: Yimin Peng <yiminp@codeaurora.org>
Date: Mon, 26 Feb 2018 15:54:28 +0800
Subject: [PATCH] trigger wdog bite for panic on gvm.

- vplatform need to wait 5s for dump to launch on host machine.
- The deferable pet timer is not suitable for vplatform on which
  vCPU thread could enter WFI for over 1 second.

Change-Id: If5d695af44838326203ced3b553c74843a0c52ff
Signed-off-by: Yimin Peng <yiminp@codeaurora.org>
---
 drivers/soc/qcom/watchdog_v2.c | 12 ++++++++++++
 kernel/panic.c                 |  6 ++++++
 2 files changed, 18 insertions(+)

diff --git a/drivers/soc/qcom/watchdog_v2.c b/drivers/soc/qcom/watchdog_v2.c
index 3528cb0..e1e783f 100644
--- a/drivers/soc/qcom/watchdog_v2.c
+++ b/drivers/soc/qcom/watchdog_v2.c
@@ -472,6 +472,13 @@ void msm_trigger_wdog_bite(void)
 	if (!wdog_data)
 		return;
 	pr_info("Causing a watchdog bite!");
+#ifdef CONFIG_MSM_GVM_QUIN
+	__raw_writel(1*WDT_HZ, wdog_data->base + WDT0_BARK_TIME);
+	__raw_writel(1*WDT_HZ, wdog_data->base + WDT0_BITE_TIME);
+	__raw_writel(1, wdog_data->base + WDT0_RST);
+	mb();
+	mdelay(5000);
+#else
 	__raw_writel(1, wdog_data->base + WDT0_BITE_TIME);
 	mb();
 	__raw_writel(1, wdog_data->base + WDT0_RST);
@@ -483,6 +490,7 @@ void msm_trigger_wdog_bite(void)
 		__raw_readl(wdog_data->base + WDT0_EN),
 		__raw_readl(wdog_data->base + WDT0_BARK_TIME),
 		__raw_readl(wdog_data->base + WDT0_BITE_TIME));
+#endif
 }
 
 static irqreturn_t wdog_bark_handler(int irq, void *dev_id)
@@ -716,7 +724,11 @@ static void init_watchdog_data(struct msm_watchdog_data *wdog_dd)
 	wdog_dd->user_pet_complete = true;
 	wdog_dd->user_pet_enabled = false;
 	wake_up_process(wdog_dd->watchdog_task);
+#ifdef CONFIG_MSM_GVM_QUIN
+	init_timer(&wdog_dd->pet_timer);
+#else
 	init_timer_deferrable(&wdog_dd->pet_timer);
+#endif
 	wdog_dd->pet_timer.data = (unsigned long)wdog_dd;
 	wdog_dd->pet_timer.function = pet_task_wakeup;
 	wdog_dd->pet_timer.expires = jiffies + delay_time;
diff --git a/kernel/panic.c b/kernel/panic.c
index 75f564a..a5ab282 100644
--- a/kernel/panic.c
+++ b/kernel/panic.c
@@ -47,6 +47,9 @@ ATOMIC_NOTIFIER_HEAD(panic_notifier_list);
 
 EXPORT_SYMBOL(panic_notifier_list);
 
+#ifdef CONFIG_MSM_GVM_QUIN
+extern void msm_trigger_wdog_bite(void);
+#endif
 static long no_blink(int state)
 {
 	return 0;
@@ -188,6 +191,9 @@ void panic(const char *fmt, ...)
 
 	trace_kernel_panic_late(0);
 
+#ifdef CONFIG_MSM_GVM_QUIN
+msm_trigger_wdog_bite();
+#endif
 	if (panic_timeout != 0) {
 		/*
 		 * This will not be a clean reboot, with everything
-- 
1.9.1

